### PR TITLE
Acp 77 local network

### PR DESCRIPTION
**Pull request by felipemadero**
## Why this should be merged
- adds local network deploy of sov blockchains
- removes usage boostrap snapshots (keeps the snapshots for CLI backwards compat)
- enables configurable number of local network nodes (sets new default to 2 nodes)
- adds teleporter deploy to blockchain deploy cmd (any network)
- adds relayer deploy to blockchain desploy cmd (any network)
- ensure same avago + relayer for network restarts in case a change is not indicated

## How this works
https://asciinema.org/a/gypjn0s5vFL3crm3afHQICqab

### CODE REVIEW

File: cmd/blockchaincmd/deploy.go [hunk 1]
```diff
@@ -990,16 +1017,98 @@ func deployBlockchain(cmd *cobra.Command, args []string) error {
...
+		if network.Kind != models.Local {
+			if flag := cmd.Flags().Lookup(skipRelayerFlagName); flag != nil && !flag.Changed {
+				ux.Logger.PrintToUser("")
+				yes, err := app.Prompt.CaptureYesNo("Do you want to use set up a local interchain relayer?")
+				if err != nil {
+					return err
+				}
+				icmSpec.SkipRelayerDeploy = !yes
+			}
+		}
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply cmd/blockchaincmd/deploy.go:1080]: the relayer is meant to be deployed automatically for local networks, and optionally for other networks.

[the remark this answers is not in the document]
Comment (felipemadero) [reply cmd/blockchaincmd/deploy.go:1080]: in the case of local machines the relayer needs more work, as it is currently not working due to 
the extra peers issue (relayer not supporting extra peers yet, there are branchs on relayer for how to fix
this but not merged/released yet)

[the remark this answers is not in the document]
Comment (felipemadero) [reply cmd/blockchaincmd/deploy.go:1080]: so relayer is probably only going to work for non sovereign settings. maybe it is best for the moment
on to just default to no (for local machines)
no?

File: cmd/interchaincmd/relayercmd/start.go [hunk 2]
```diff
@@ -50,7 +57,16 @@ func start(_ *cobra.Command, _ []string) error {
 		return err
 	}
 	switch {
-	case network.Kind == models.Local || network.Kind == models.Fuji:
+	case network.ClusterName != "":
```
Comment (arturrez) [inline cmd/interchaincmd/relayercmd/start.go:60]: is this check is enough to decide if cluster is remote and ssh needed?
Do we need to also load ClusterConfig and check `clusterConf.Local` ?

Comment (arturrez) [reply cmd/interchaincmd/relayercmd/start.go:60]: pls ignore this comment. This is code that was tested already and just moved  within code

File: cmd/networkcmd/start.go [hunk 3]
```diff
@@ -42,30 +52,39 @@ By default, the command loads the default snapshot. If you provide the --snapsho
...
 	}
 
-	cmd.Flags().StringVar(&userProvidedAvagoVersion, "avalanchego-version", latest, "use this version of avalanchego (ex: v1.17.12)")
-	cmd.Flags().StringVar(&avagoBinaryPath, "avalanchego-path", "", "use this avalanchego binary path")
-	cmd.Flags().StringVar(&snapshotName, "snapshot-name", constants.DefaultSnapshotName, "name of snapshot to use to start the network from")
+	cmd.Flags().StringVar(&startFlags.UserProvidedAvagoVersion, "avalanchego-version", latest, "use this version of avalanchego (ex: v1.17.12)")
+	cmd.Flags().StringVar(&startFlags.AvagoBinaryPath, "avalanchego-path", "", "use this avalanchego binary path")
+	cmd.Flags().StringVar(&startFlags.RelayerBinaryPath, "relayer-path", "", "use this relayer binary path")
+	cmd.Flags().StringVar(&startFlags.SnapshotName, "snapshot-name", constants.DefaultSnapshotName, "name of snapshot to use to start the network from")
+	cmd.Flags().Uint32Var(&startFlags.NumNodes, "num-nodes", constants.NumDefaultLocalNetworkNodes, "number of nodes to be created on local network")
```
Comment (arturrez) [inline cmd/networkcmd/start.go:63]: nit: `NumDefaultLocalNetworkNodes` should be `LocalNetworkNumNodes`

File: cmd/networkcmd/start.go [hunk 4]
```diff
@@ -95,154 +109,184 @@ func StartNetwork(*cobra.Command, []string) error {
...
+	}
+	if flags.SnapshotName == "" {
+		flags.SnapshotName = constants.DefaultSnapshotName
 	}
 
-	ux.Logger.PrintToUser("Node logs directory: %s/node<i>/logs", resp.ClusterInfo.LogRootDir)
-	ux.Logger.PrintToUser("")
-	ux.Logger.PrintToUser("Network ready to use.")
-	ux.Logger.PrintToUser("")
+	snapshotPath := filepath.Join(app.GetSnapshotsDir(), "anr-snapshot-"+flags.SnapshotName)
```
Comment (arturrez) [inline cmd/networkcmd/start.go:146]: nit: we use "anr-snapshot-" prefix in multiple places. Should we have a single func that returns snapshotPath for the name?

File: cmd/networkcmd/start.go [hunk 5]
```diff
@@ -95,154 +109,184 @@ func StartNetwork(*cobra.Command, []string) error {
...
+	if utils.DirectoryExists(snapshotPath) {
+		var startMsg string
+		if flags.SnapshotName == constants.DefaultSnapshotName {
```
Comment (arturrez) [inline cmd/networkcmd/start.go:149]: nit: not worth displaying snapshot name for the user if it was provided by the user anyway.
so we don't need this if else statement

File: cmd/networkcmd/upgrade.json [hunk 6]
```diff
@@ -0,0 +1,16 @@
+{
```
Comment (arturrez) [inline cmd/networkcmd/upgrade.json:1]: this is the same file as `pkg/constants/etnaDevnet/upgrade.json` no need to create another file here

**Review by arturrez — APPROVED** [review]

File: cmd/blockchaincmd/deploy.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (felipemadero) [reply cmd/blockchaincmd/deploy.go]: latest relayer is good enough to work on etna, but for fuji needs one more detail (allowing private ips....)
so, defaulting to yes for local machines is general, but to no for fuji

**Comment by felipemadero** [comment]
> ```
> AVL-7H2W7V:avalanche-cli raymondsukanto$ ./bin/avalanche subnet deploy newSubnet2
> ✔ Yes
> ✔ Subnet-EVM
> ✔ Proof Of Authority
> ✔ Get address from an existing stored key (created from avalanche key create or avalanche key import)
> ✔ ewoq
> ✓ Validator Manager Contract owner address 0x8db97C7cEcE249c2b98bDC0226Cc4C2A57BF52FC
> ✔ I want to use defaults for a test environment
> Chain ID: 123213
> Token Symbol: AVAX
> Token Symbol: AVAX
> prefunding address 0x8db97C7cEcE249c2b98bDC0226Cc4C2A57BF52FC with balance 1000000000000000000000000
> File /Users/raymondsukanto/.avalanche-cli/subnets/newSubnet2/chain.json successfully written
> ✓ Successfully created blockchain configuration
> Run 'avalanche blockchain describe' to view all created addresses and what their roles are
> ✔ Local Network
> Deploying [newSubnet2] to Local Network
> 
> Backend controller started, pid: 81891, output at: /Users/raymondsukanto/.avalanche-cli/runs/server_20241127_120947/avalanche-cli-backend.log
> 
> Installing avalanchego-v1.12.0...
> Error: failed setting up local environment: unable to download binary: unexpected http status code: 404
> ```
> 
> Need to change to using latest avalanche go pre release version for local network. Seems that its trying to use avalanchego-v1.12.0 instead of avalanchego-v1.12.0-fuji

this happened when local machine cluster is up. strange old code tried to get cluster avalanchego node version,
avalanchego answers avalanchego-v1.12.0 instead of avalanchego-v1.12.0-fuji.

fixed all logic related to obtaining avalanchego version. also, if cluster is up, as currently there is incompatibility
between local network and local clusters due to shared endpoint (a PR that will come in the future to fix the conflict)
local network will fail if local cluster is up

**Comment by felipemadero** [comment]
> awm-relayer/v1.5.1

latest relayer release awm-relayer/v1.5.1 broke assumption regarding relayer versions (previously, just v1.5.1).
fixed code to support new version format

**Comment by felipemadero** [comment]
> also seems that we are only spinning up 2 nodes on local network? Why 2 nodes?

2 nodes is the minimum required to have a local network with healthy consensus. 
also, takes much less bootstrapping time, making all development faster.
also, takes much less resources (eg disk/network/cpu), making it cheaper to be
used in environments such as avacloud devnets

File: cmd/networkcmd/upgrade.json (hunk 6 above)
Comment (felipemadero) [reply cmd/networkcmd/upgrade.json:1]: I believe etna one is going to be removed soon, so will keep this one

**Pull request merged**